### PR TITLE
Allow pipeline writer to spit out Ogg directly, including when seeking

### PR DESCRIPTION
**Pull request by philippe44**
Use ogg v 0.8 to use directly inner writer instead of the complicated workaround in previous patch https://github.com/librespot-org/librespot/pull/560

### CODE REVIEW

**Comment by sashahilton00** [comment]
I've merged a fix for the CI, and updated the Cargo.lock file, along with bumping the minimum rust version to `1.46.0`, since the current rust version is `1.49.0`, and `vec_drain_as_slice` was introduced in `1.46.0` and thus was causing issues during compilation.

**Comment by philippe44** [comment]
Thanks!

**Comment by ashthespy** [comment]
> along with bumping the minimum rust version to `1.46.0`, since the current rust version is `1.49.0`, and `vec_drain_as_slice` was introduced in `1.46.0` and thus was causing issues during compilation.

Shouldn't this then be a feature flag gated feature? 
For the core library, it would be nice to maintain Debian's current stable (buster) version of Rust as the MSRV (1.41).

**Comment by sashahilton00** [comment]
@ashthespy probably not a bad idea. I personally don't care about keeping the supported compiler version low, but I can see the argument for it. @philippe44 are you ok to add in a compile time flag for this feature and drop the CI MSRV back to `1.41.0`?

**Comment by philippe44** [comment]
I'll have to look at how to make that a clean compile time option (I'm a Rust noob, so everything requires learning) or I'll see if I can use another way to get the data instead of using vec_drain_as_slice

**Comment by plietar** [comment]
Isn’t that code just equivalent to `vec![data[0] as i16]`?

**Comment by plietar** [comment]
I kind of worry about passing the OGG data as i16s. For instance in this particular line, an extraneous byte gets inserted in the u8 -> i16 -> u8 conversion process.

I think it would be more appropriate to turn AudioPacket into a enum with a Vec<i16> entry for PCM samples and a Vec<u8> for OGG data.

The pipe backend matches on the packet type and does the right thing. Other backends crash on OGG data.

**Comment by philippe44** [comment]
> Isn’t that code just equivalent to `vec![data[0] as i16]`?
>
I'm still a Rust noob and I had a hard time with the first PR that forced me to use Ref/RefCell/Arc/interior mutation :-(

**Comment by philippe44** [comment]
> I kind of worry about passing the OGG data as i16s. For instance in this particular line, an extraneous byte gets inserted in the u8 -> i16 -> u8 conversion process.
> 
> I think it would be more appropriate to turn AudioPacket into a enum with a Vec entry for PCM samples and a Vec for OGG data.
> 
I agree but this is always the last byte of a stream so it will be dropped. I did that b/c I did not want to have even more change to the rest of the code. I can study that option if you prefer
> The pipe backend matches on the packet type and does the right thing. Other backends crash on OGG data.

**Comment by philippe44** [comment]
Looking at the pipe.rs code, it seems that it expects Vec<i16>
```
impl Sink for StdoutSink {
    fn start(&mut self) -> io::Result<()> {
        Ok(())
    }

    fn stop(&mut self) -> io::Result<()> {
        Ok(())
    }

    fn write(&mut self, data: &[i16]) -> io::Result<()> {
        let data: &[u8] = unsafe {
            slice::from_raw_parts(
                data.as_ptr() as *const u8,
                data.len() * mem::size_of::<i16>(),
            )
        };

        self.0.write_all(data)?;
        self.0.flush()?;

        Ok(())
    }
}
```
I understand I can modify that as well, but how much are the maintainers willing to let such modification creep too far in the code?

**Comment by sashahilton00** [comment]
I was just reading up on rust enums, they are pretty elegant. However, in the interests of expediency do we want to merge this and evaluate the audio packet changes in a separate PR - such a change looks like it will require some rewriting of the existing back ends.

**Comment by philippe44** [comment]
I think I have been able to use enum properly, well at least it works, produce the expected result and does not have too many ripple effects. The only one is the addition of a write_bytes() trait that has a default version that does nothing.

I've done per asked, a samples() function to access i16 and a bytes() function to access raw data. The passthrough is now properly writing u8 and not adding (in some case) an extra byte.

As said before, excuse Rust rookie style & practices mistakes. I also have an issue with the merger that I'm not sure I now how to fix. My cargo.lock should never have been pushed and the audio/Cargo.toml should just add ogg, I don't know why it is in conflict now

**Comment by philippe44** [comment]
I don't think I can fix the --locked issue

**Comment by philippe44** [comment]
Any love/hate/opinion for that version?

**Comment by sashahilton00** [comment]
I'll try and fix the cargo.lock file tomorrow. There are a few PRs which are having problems due to older rust versions, to the point where it may make sense to just bump the MSRV to `1.42.0`. With regards to the enums, @plietar is the enum implementation what you were expecting? Also, given the change to AudioPacket which is used in various backends, have they been tested to ensure that they are functioning as expected?

**Comment by plietar** [comment]
> With regards to the enums, @plietar is the enum implementation what you were expecting?

Generally looks good. I'll have a proper look through the code this weekend.

**Comment by sashahilton00** [comment]
We should also update the instructions/dependencies in `COMPILING.md` wrt the Jack backend now using Rodio

**Comment by sashahilton00** [comment]
Are we good to merge this?

**Review by plietar — COMMENTED** [review]
Could you add a high level description of how the passthru decoder works please.
In particular what is the purpose of the absgp stitching, and how does seeking work.

On every seek we re-send the same headers? Is that something decoders tolerate / require?

File: audio/src/lib.rs
Comment (Johannesd3) [reply audio/src/lib.rs:42]: If you go with pattern matching instead of using this functions (as @plietar suggested below), you could also remove this functions entirely. At least I wouldn't expect such harmless-looking functions to panic.

File: audio/src/lib.rs
Comment (philippe44) [reply audio/src/lib.rs:62]: It's to simplify the usage in playback/src/player.rs where it's convenient to access len prior knowing the type of data underneath (handle_packet)

File: src/main.rs
Comment (Johannesd3) [inline src/main.rs:211]: I might overlook something, but shouldn't we check that the backend is indeed pipe if this option is present?

File: playback/src/audio_backend/mod.rs
Comment (philippe44) [reply playback/src/audio_backend/mod.rs:13]: I agree but then I would have to modify every existing backend which I would assume was not desirable, that's the reason I ended up with this. I did not want that PR to be a virus spreading across code.

File: playback/src/player.rs
Comment (Johannesd3) [inline playback/src/player.rs:381]: There were already previously two different `VorbisDecoder`s, and it was solved with feature flags. Now, a third decoder is added, and a trait is used instead. It would be more consistent IMHO to use either feature flags or traits for all decoders. A third option (with less overhead than boxed traits) would be an `enum`, which is sometimes an alternative to traits if there's a well-known number of implementors.

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs:184]: will do

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs:181]: It should not, it was just a sanity check but that can be avoided. I'll remove it

File: playback/src/audio_backend/mod.rs
Comment (Johannesd3) [reply playback/src/audio_backend/mod.rs:13]: Would it be an option to add an optional type parameter to `Sink`? Like
```rust
pub trait Sink<S = i16> {
    fn write(&mut self, data: &[S]) -> io::Result<()>; 
}
```

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs:77]: It is a static because it survives decoder instances and it is stitching together the global timeline. If not, multiple tracks will have an incorrect position and the resulting file will not be readable

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs:5]: done

File: audio/src/lib.rs
Comment (philippe44) [reply audio/src/lib.rs:42]: @Johannesd3 - understood but back to my comment, then I would need to modify all backends to do the matching and I was trying to avoid that like plague

File: audio/src/lib.rs
Comment (philippe44) [reply audio/src/lib.rs:37]: I was thinking that would more describe a blob so that if one day there is another codec, we could still refer to the outcome as 'bytes'

File: src/main.rs
Comment (philippe44) [reply src/main.rs:211]: Or, we could even force the backend to be pipe - that was my initial opinion

File: audio/src/lib.rs
Comment (plietar) [reply audio/src/lib.rs:42]: @Johannesd3 I actually think these functions are useful inside the backends that only support one kind of packet. Of course each backend could also match and do the panic themselves, but it's cleaner in a version.

File: playback/src/audio_backend/mod.rs
Comment (plietar) [reply playback/src/audio_backend/mod.rs:13]: > I would have to modify every existing backend which I would assume was not desirable

Modifying every backend is perfectly fine if the overall result is nicer.

-----

@Johannesd3 I'm not sure this actually works. The player holds a `Box<dyn Sink>`, so it is important that all sinks expose the same interface.

Comment (Johannesd3) [reply playback/src/audio_backend/mod.rs:13]: It works for certain, but not without modifying very much existing code, so probably it was a bad idea.

File: src/main.rs
Comment (plietar) [reply src/main.rs:211]: I prefer an error than forcing the backend. It would be less surprising to users.

File: playback/src/player.rs
Comment (plietar) [reply playback/src/player.rs:381]: Feature flags don't work because we need to switch between the two based on runtime flags.
I don't really see a usecase for switching between libogg and lewton at runtime, so feature flags is fine there.

I agree there is a bit of inconsistency, but it seems reasonable and actually simpler this way.

I think the overhead of boxed trait is completely fine in this context.

File: audio/src/lib.rs
Comment (plietar) [reply audio/src/lib.rs:62]: Hmm, would an `empty(&self) -> bool` be enough in that case?

File: audio/src/passthrough_decoder.rs
Comment (plietar) [reply audio/src/passthrough_decoder.rs:77]: Ah right. So is the idea that for the lifetime of the process the granule only ever gets incremented?

Comment (Johannesd3) [reply audio/src/passthrough_decoder.rs:77]: I don't understand what apsgp does, so I might be wrong. But librespot could be used as library, and multiple PassthroughDecoders could be created at the same time for whatever reason. What would happen then? There isn't any protection against concurrent access to `BASEGP_PAGE`, is it?

Comment (philippe44) [reply audio/src/passthrough_decoder.rs:77]: > Ah right. So is the idea that for the lifetime of the process the granule only ever gets bigger?
>
Yes

File: audio/src/lib.rs
Comment (philippe44) [reply audio/src/lib.rs:62]: you're torturing me but ok :-)

File: audio/src/lib.rs
Comment (philippe44) [reply audio/src/lib.rs:42]: Done

File: audio/src/libvorbis_decoder.rs
Comment (philippe44) [reply audio/src/libvorbis_decoder.rs:76]: Here I'm struggling ... I thought it would be define in libvorbis because it does that for AudioDecoder. And I can't find a proper option if I define it lib.rs (help!)

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs:77]: > I don't understand what apsgp does, so I might be wrong. But librespot could be used as library, and multiple PassthroughDecoders could be created at the same time for whatever reason. What would happen then? There isn't any protection against concurrent access to `BASEGP_PAGE`, is it?
>
There is probably no perfect answer. This relates to granule position's page. It is used to seek into file as well as a way to know the duration of an ogg file:  the granule of the last page of the file give you the duration, there is no other way. So if I don't have that incremented properly, then duration is wrong and players can be totally confused. There is no perfect solution anyway b/c when a seek is performed, the granule will overshoot, but that's better than undershoot that lead to premature ending. Re having that as a global, I agree and I don't like globals, but otherwise I would need to offer a "context" which is much more complications and non-compatibilty with existing code. I could have created another new() with an extra parameter to reset it but I think Rust does not offer overloading/mangling

File: playback/src/audio_backend/mod.rs
Comment (philippe44) [reply playback/src/audio_backend/mod.rs:13]: > > I would have to modify every existing backend which I would assume was not desirable
> 
> Modifying every backend is perfectly fine if the overall result is nicer.
> 
So... are you really sure you want me to do that?

File: playback/src/player.rs
Comment (philippe44) [reply playback/src/player.rs:1080]: ok, but I think I need something like
```
	if let AudioPacket::Samples(ref mut data) = packet {
						if let Some(ref editor) = self.audio_filter {
							editor.modify_stream(data)
						}
 
						if self.config.normalisation && normalisation_factor != 1.0 {
							for x in data.iter_mut() {
								*x = (*x as f32 * normalisation_factor) as i16;
							}
						}
```
but I'm not sure, Rust is still giving me headaches at this point

File: src/main.rs
Comment (philippe44) [reply src/main.rs:211]: ok

**Comment by philippe44** [comment]
> Could you add a high level description of how the passthru decoder works please.
> In particular what is the purpose of the absgp stitching, and how does seeking work.
> 
> On every seek we re-send the same headers? Is that something decoders tolerate / require?
>
I had to review things properly after that discussion. I was struggling too much with Rust, using Ogg backend at the same time. So there was a fair bit of remaining issues that are hopefully corrected now in addition of complying to most of your requirements.

A a general comment, this patch is to help the (spotty)[https://github.com/michaelherger/spotty]  project where the main objective is to stream *only* one track per instance (with and optional starting offset) then launch another instance for the next track.

The passthru decoder nomally simply copies the entire stream to stdout, untouched. Now, there are a few problems because OggVorbis is not a fully streamable format as there are 3 headers (ident, comment and setup) per stream and they are all required especially the setup contains information to configure the vorbis decoder. Streams are divided into pages that have an own sequence number. There is no global header that tells the duration, instead "granule" system tells the position of each page and seek has to deal with that. Players get the whole duration of a file by reading the granule of the last page of the file.

1- When seeking within one track, headers must be added *once* at the beginning of the new position, so they must be gathered from the initial stream and then the seek must be done. Seek in librespot is done (AFAIK) by streaming the actual full data until the expected offset. I initially left some basic errors there as I misunderstood the way librespot chain tracks (see issue below) and there is no need to repeat the header when seeking, but the granule must always increment otherwise many players will terminate early. I don't think there is a perfect solution for granule handling when seeking (as one can seek forward and backward and this *not* a new instance of the decoder), so I've opted for a granule that always increases so that at least the player does not fall short. I'll continue thinking if there is a better but reasonably easy solution. 

2- When chaining tracks (without user pressing next), I initially thought that librespot was continuing with the same decoder instance, but it's not the case. A new decoder is created for every track, even when they flow "naturally" but so I have to keep a global context of granule if I want an incrementing value across tracks (per problem described above). I know when the last page/granule of a given stream is received as I have a last packet flag. Still, I cannot assume that at some point I won't receive a continuous data flow with multiple streams.

3- Last, when user changes tracks, librespot also creates a new decoder instance but of course in that case there is no last page so I need a way to continue this granule incrementation thingy.

Hope this helps and I'll continue thinking about options, mean while I'll push the updates I did

File: audio/src/passthrough_decoder.rs
Comment (Johannesd3) [reply audio/src/passthrough_decoder.rs:65]: Ah, it's clear for me now.

But using `unsafe` in a library such that undefined behaviour could occur if you use it wrong should definitely be avoided.

You could use 
```rust

static BASEGP_PAGE: AtomicU64::new(0);

// Adding
BASEGP_PAGE.fetch_add(n, Ordering::Relaxed);
```

Even though it doesn't make sense to use multiple instance at the same time, it's at least no UB.

File: audio/src/lib.rs
Comment (Johannesd3) [inline audio/src/lib.rs:67]: AFAIK, `is_empty` would be more usual in Rust. Vecs also have a function `is_empty` which you could use here. But that's bikeshedding.

**Review by Johannesd3 — CHANGES_REQUESTED** [review]

**Comment by sashahilton00** [comment]
There's quite a bit of stuff to be tweaked here before it's ready to merge just looking over the discussion. Can I propose that given that this is not a breaking change we aim to get this in the `0.2.1` release? I'm conscious that the `librespot-tremor` issue for newer versions of rust is causing some problems for Spotifyd and would like to release a new version sooner rather than later.

**Comment by philippe44** [comment]
> There's quite a bit of stuff to be tweaked here before it's ready to merge just looking over the discussion. Can I propose we aim to get this in the `0.2.1` release? I'm conscious that the `librespot-tremor` issue for newer versions of rust is causing some problems for Spotifyd and would like to release a new version sooner rather than later.
>
You are the maintainer, I'll do as you recommend. My only ask is to not lose "momentum" on the discussion

File: audio/src/lib.rs
Comment (philippe44) [reply audio/src/lib.rs]: done

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs]: done

**Comment by philippe44** [comment]
I've addressed all the comments I could but I'm not sure why I have this conflict on player.rs. Let me know if you want to restart a fresh PR or stash things together

**Comment by Johannesd3** [comment]
> There's quite a bit of stuff to be tweaked here before it's ready to merge just looking over the discussion. Can I propose we aim to get this in the `0.2.1` release? I'm conscious that the `librespot-tremor` issue for newer versions of rust is causing some problems for Spotifyd and would like to release a new version sooner rather than later.

@sashahilton00 Maybe apply/cherry-pick the tremor fix onto the current 0.1.3 and make a 0.1.4 of it?

**Comment by philippe44** [comment]
I've finally cleaned up (I hope for the best) the different commits and rebased to the latest dev to make differences easier to follow

File: playback/src/player.rs
Comment (philippe44) [inline playback/src/player.rs:737]: I'm not sure this is the best option but my Rust "rookieness" really hurts here

Comment (Johannesd3) [reply playback/src/player.rs:737]: I think it's fine. You _could_ use `Result::map` and  `map_err` instead of `match`, but I don't think it's a big win.

Comment (philippe44) [reply playback/src/player.rs:737]: thx!

File: audio/src/lib.rs
Comment (Johannesd3) [reply audio/src/lib.rs]: Would you please also replace the `.len() == 0`? [Clippy would give a warning](https://rust-lang.github.io/rust-clippy/master/#len_zero) (and we have already enough of those warnings imo...)

File: audio/src/passthrough_decoder.rs
Comment (Johannesd3) [inline audio/src/passthrough_decoder.rs:142]: Use the return value instead of loading it again below. Atomic operations are expensive.

File: audio/src/passthrough_decoder.rs
Comment (Johannesd3) [inline audio/src/passthrough_decoder.rs:139]: Feels a bit hacky. Maybe `Option<u64>` would be clearer? And maybe rename the field to `last_absgp_page`?

**Review by Johannesd3 — CHANGES_REQUESTED** [review]

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs:68]: Done

File: audio/src/passthrough_decoder.rs
Comment (Johannesd3) [reply audio/src/passthrough_decoder.rs:162]: It doesn't seem removed.

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs]: Done

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs]: Done but what do you expect from a C developer :-) ?

File: audio/src/passthrough_decoder.rs
Comment (Johannesd3) [inline audio/src/passthrough_decoder.rs:140]: You could use something like 
```rust
let lastgp_page = *self.lastgp_page.get_or_insert(pckgp_page);
```

to avoid the `unwrap`s below.

**Comment by Johannesd3** [comment]
I have no objections anymore. While the static granule seems not ideal, I don't know a better solution either.

**Comment by philippe44** [comment]
> I have no objections anymore. While the static granule seems not ideal, I don't know a better solution either.
>
Thank you for your comments - Rust is a steep learning curve :-)

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs:160]: oh this one, no it's on purpose. You might have nothing produced by the packet writer

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs]: done

**Comment by Johannesd3** [comment]
I don't know much about ogg, so this idea might be stupid: Create some kind of OggSink and keep track of the granule in there. The sink would get the vorbis data along with the granule differences. Would it work? However, this should not block this PR.

File: audio/src/lib.rs
Comment (sashahilton00) [reply audio/src/lib.rs:37]: I think `OggData` would be more self explanatory. If another codec is implemented at a later date, then we can think about renaming this, but for now Spotify encodes everything with `ogg` and doesn't look to be changing anytime soon.

File: audio/src/libvorbis_decoder.rs
Comment (sashahilton00) [reply audio/src/libvorbis_decoder.rs:76]: Unless I'm missing something, it should be possible to define

```
#[cfg(feature = "with-tremor")]
extern crate librespot_tremor as vorbis;
#[cfg(not(feature = "with-tremor"))]
extern crate vorbis;
```
at the top of `lib.rs`, and then further down by `AudioError`, convert from the `VorbisError` using a `From` implementation:

```
impl From<vorbis::VorbisError> for AudioError {
    fn from(err: vorbis::VorbisError) -> AudioError {
        AudioError::VorbisError(VorbisError::from(err))
    }
}
```
And then remove the existing implementations from the `libvorbis`/`lewton` decoders

File: playback/src/audio_backend/mod.rs
Comment (sashahilton00) [reply playback/src/audio_backend/mod.rs:13]: I think it would be cleaner. Matching from each backend doesn't really add any code complexity, and it makes more sense than multiple methods to read/write the same underlying data, albeit in a different format.

File: audio/src/lib.rs
Comment (philippe44) [reply audio/src/lib.rs]: Done

File: playback/src/audio_backend/mod.rs
Comment (philippe44) [reply playback/src/audio_backend/mod.rs]: done

**Comment by philippe44** [comment]
So I think I've made all the requested changes, including in all backends + error management. I hope this is a better "Rust-compliant".
I've also made modification in the granule management because I was wrong, the granule does not need to be monotonic across chained streams, it only needs to be within the same stream (when seeking). Still, it was not working before because I did not realize that spotify does not set the stream serial number and in that case, the Ogg stream seems to be a single one and there, a monotonic granule was required. I've just added a fake stream serial if it is set to 0

**Comment by Johannesd3** [comment]
It's great that the problem with the global granule is solved, thanks!

Maybe you could run `cargo clippy` and fix the warnings that are related to this PR?

File: audio/src/passthrough_decoder.rs
Comment (Johannesd3) [inline audio/src/passthrough_decoder.rs:177]: I think the `IOError` case will never occur, but I won't guarantee it: This variant is used only if `PacketWriter<Vec<u8>>::write_packet` fails. Looking at the source code, it seems as if this function only passes through the errors of the underlying `Write` object. But writing to a `Vec` never fails, so you could as well `unwrap` the `Result`s of `write_packet`. (Is there anyone who could confirm this?)

If this variant is removed, you could also remove the `NotVorbisError` variant and replace it by `OggReadError::InvalidData`. So there wouldn't be a need for an enum anymore, you could simply use a "newtype" struct as in the vorbis decoders.

File: audio/src/lib.rs
Comment (Johannesd3) [inline audio/src/lib.rs:73]: Is it the right way to make `AudioDecoder` a trait, but `AudioError` an enum?

**Review by Johannesd3 — COMMENTED** [review]
I added a few comments regarding error handling. 

@sashahilton00 @ashthespy Do you think this would be an appropriate place to start using `thiserror`? It would spare us a lot of boilerplate here.

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs:177]: yes, more simplification :-)

File: audio/src/lib.rs
Comment (philippe44) [reply audio/src/lib.rs:73]: done then

File: audio/src/passthrough_decoder.rs
Comment (philippe44) [reply audio/src/passthrough_decoder.rs:193]: This is the same as the others (lewton & libvorbis)

Comment (Johannesd3) [reply audio/src/passthrough_decoder.rs:193]: No, there it's `self.0`, so it is forwarded to the inner type. Here, it's forwarded to itself.

Comment (philippe44) [reply audio/src/passthrough_decoder.rs:193]: argh....

**Comment by philippe44** [comment]
Let me know if there is something else expected from me

**Comment by sashahilton00** [comment]
Just released `0.1.5` with the tremor fix, so there isn't pressure on this PR anymore. Will wait to hear from anyone that has any further thoughts/suggestions, otherwise this looks pretty much ready to go @philippe44

**Comment by sashahilton00** [comment]
@philippe44 would you mind squashing the commits?

**Comment by Johannesd3** [comment]
May I point out that the commit has a not very meaningful name?

**Comment by sashahilton00** [comment]
If you could rename the commit to something more meaningful then that would be good. After that I'll merge it.

**Comment by Johannesd3** [comment]
@ashthespy (or @sashahilton00) After this is merged, could you try merging `dev` into `tokio_migration`, if you can spare the time? The more they diverge, the harder it will be to merge eventually.

**Comment by sashahilton00** [comment]
Merged. Please update the wiki accordingly as requested above.

**Comment by philippe44** [comment]
> Merged. Please update the wiki accordingly as requested above.
>
Done

**Pull request merged**
